File: vanilla_call/julia/explicit_method.jl
using DataFrames

# Set up parameters.
r     = 0.03;       # risk-free rate
q     = 0;          # dividend yield
sigma = 0.2;        # volatility of underlying
T     = 10;         # horizon (right edge)
Smin  = 0;          # top edge
Smax  = 200;        # bottom edge
Nt    = 10^5+1;     # resolution in time
Nx    = 200+1;      # resolution in space
dt    = T/(Nt-1);
dx    = (Smax-Smin)/(Nx-1);
# Call with strike K
K     = 105;
# Set up grids
x_grid = linspace(Smin,Smax,Nx);
t_grid = linspace(0,T,Nt);
# Set up initial conditions
initial_conditions = map(z -> max(0,z-K),x_grid);
# Set up Dirichlet boundary conditions
boundary_bottom    = zeros(Nt);
boundary_top       = max(0,Smax-K)*map(exp,-r*t_grid);

# Explicit method
#   Simple tridiagonal forward operator A.
a_e = 0.5*[i^2 for i in linspace(0,Nx-1,Nx)]*(sigma^2)*dt/(dx^2);
b_e = 0.5*(r-q)*[i for i in linspace(0,Nx-1,Nx)]*dt/dx;
l_e = a_e-b_e; d_e = 1-2*a_e-r*dt; u_e=a_e+b_e;
# Efficiently store sparse tridiagonal matrix
explicit = spdiagm(0 => d_e, -1 => l_e[2:Nx], 1 => u_e[1:Nx-1]);
V = zeros((Nx,Nt));
V[:,1] = initial_conditions;

# @time for j in 2:Nt
for j in 2:Nt
    Vk          = copy(V[:,j-1]);
    V[:,j]      = copy(explicit*Vk);            # Forward operator
    V[1,j]     += l_e[1]*boundary_bottom[j];    # Boundary conditions
    V[Nx,j]    += u_e[Nx]*boundary_top[j];      # Boundary conditions
end
# 0.618032 seconds (2.03 M allocations: 717.108 MiB, 43.11% gc time)

using Seaborn
ioff()
Seaborn.plot(V[:,[1,20001,40001,60001,80001,100001]])
Seaborn.savefig("explicit_method.png",dpi=300,transparent=true,bbox_inches="tight")

File: vanilla_call/python/explicit_method.py
'''
    PDE is
    -DtV + (r-q)*x*DxV + 0.5*sigma^2*x^2*DxxV - r*V = 0
'''

import numpy as np
import pandas as pd
from scipy.sparse import diags
import matplotlib as mpl
import matplotlib.pyplot as plt

'''
    Set up parameters.
'''

r     = 0.03        # risk-free rate
q     = 0           # dividend yield
sigma = 0.2         # volatility of underlying
T     = 10          # horizon (right edge)
Smin  = 0           # top edge
Smax  = 200         # bottom edge
Nt    = 10**5+1     # resolution in time
Nx    = 200+1       # resolution in space
dt    = T/(Nt-1)
dx    = (Smax-Smin)/(Nx-1)
# Call with strike K
K     = 105
# Set up grids
x_grid = np.linspace(Smin,Smax,Nx)
t_grid = np.linspace(0,T,Nt)
# Set up initial conditions
initial_conditions = pd.Series(x_grid).apply(lambda z:max(0,z-K))
# Set up Dirichlet boundary conditions
boundary_bottom    = pd.Series(np.zeros(Nt),index=t_grid)
boundary_top       = pd.Series(max(0,Smax-K)*np.exp(-r*t_grid),index=t_grid)

'''
    Explicit method
        Simple tridiagonal forward operator A.
'''

alpha = pd.Series(0.5*np.linspace(0,Nx,Nx)**2*(sigma**2)*dt/(dx**2),index=x_grid)
beta = pd.Series(0.5*(r-q)*np.linspace(0,Nx,Nx)*dt/dx,index=x_grid)
l_e = alpha-beta ; d_e = 1-2*alpha-r*dt ; u_e = alpha+beta
# Efficiently store sparse tridiagonal matrix
explicit = diags([d_e,u_e.iloc[0:Nx-1],l_e.iloc[1:Nx] ],[0,1,-1])
V = pd.DataFrame(np.zeros((Nx,Nt)))
V.iloc[:,0] = initial_conditions

tic = time.clock()
for j in range(1,Nt):       # Approximately 302 seconds on my machine
    Vk             = V.iloc[:,j-1].copy()
    V.iloc[:,j]    = explicit*Vk                    # Forward operator
    V.iloc[0,j]    = boundary_bottom.iloc[j]        # Impose boundary conditions
    V.iloc[Nx-1,j] = boundary_top.iloc[j]           # Impose boundary conditions
    print(j)
V.index = np.linspace(Smin,Smax,Nx)
V.columns = np.linspace(0,T,Nt)
toc = time.clock()
toc-tic

import seaborn as sns
V.iloc[:,[0,20000,40000,60000,80000,100000]].plot()
plt.savefig('explicit_method.png',dpi=300,
            transparent=True,bbox_inches='tight')
plt.close("all")
